presentation/src/main/java/me/androidbox/presentation/login/viewmodel/LoginViewModel.kt

```diff
@@ -12,44 +9,50 @@ import kotlinx.coroutines.launch
 import me.androidbox.domain.authentication.ResponseState
 import me.androidbox.domain.authentication.model.Login
 import me.androidbox.domain.authentication.usecase.LoginUseCase
+import me.androidbox.presentation.login.screen.LoginScreenEvent
+import me.androidbox.presentation.login.screen.LoginScreenState
 import javax.inject.Inject
 
 @HiltViewModel
 class LoginViewModel @Inject constructor(
     private val loginUseCase: LoginUseCase
 ) : ViewModel() {
 
-    private val loginMutableState: MutableStateFlow<ResponseState<Login>?> = MutableStateFlow(null)
-    val loginState = loginMutableState.asStateFlow()
-
-    var email by mutableStateOf("")
-        private set
-
-    var password by mutableStateOf("")
-        private set
-
-    var isPasswordVisible by mutableStateOf(false)
-        private set
-
-    fun onUsernameChanged(newUsername: String) {
-        email = newUsername
-    }
-
-    fun onPasswordChanged(newPassword: String) {
-        password = newPassword
-    }
-
-    fun onPasswordVisibilityChanged() {
-        isPasswordVisible = !isPasswordVisible
+    private val loginResponseMutableState: MutableStateFlow<ResponseState<Login>?> = MutableStateFlow(null)
+    val loginResponseState = loginResponseMutableState.asStateFlow()
+
+    private val loginScreenMutableState = MutableStateFlow(LoginScreenState("", "", false))
+    val loginScreenState = loginScreenMutableState.asStateFlow()
+
+    fun onLoginEvent(loginScreenEvent: LoginScreenEvent) {
+        when(loginScreenEvent) {
+            is LoginScreenEvent.OnEmailChanged -> {
+                loginScreenMutableState.value = loginScreenState.value.copy(
+                    email = loginScreenEvent.email
+                )
+            }
+            is LoginScreenEvent.OnPasswordChanged -> {
+                loginScreenMutableState.value = loginScreenState.value.copy(
+                    password = loginScreenEvent.password
+                )
+            }
+            is LoginScreenEvent.OnPasswordVisibilityChanged -> {
+                loginScreenMutableState.value = loginScreenState.value.copy(
+                        isPasswordVisible = !loginScreenState.value.isPasswordVisible
+                    )
+            }
+            LoginScreenEvent.OnLoginUser -> {
+                loginUser(loginScreenState.value.email, loginScreenState.value.password)
+            }
+        }
     }
-
-    fun loginUser(email: String, password: String) {
+    private fun loginUser(email: String, password: String) {
         viewModelScope.launch {
-            loginMutableState.value = loginUseCase.execute(email, password)
+            loginResponseMutableState.value = loginUseCase.execute(email, password)
         }
     }
 
     fun saveCurrentUserDetails(login: Login) {
-        /* Will call use-case to save user login details */
+        /* Will call use-case to save user login details to shared preferences WIP */
     }
```

Comment:
This function doesn't need to be called from the UI. You can just call it after login was successful in the ViewModel or directly in the repo when you get the token response

Reply:
Would this be acceptable, as I would have to check to make sure the response is success first.
```
 private fun loginUser(email: String, password: String) {
        viewModelScope.launch {
            val loginResponseState = loginUseCase.execute(email, password)

            if (loginResponseState is ResponseState.Success) {
                saveCurrentUserDetails(loginResponseState.data)
            }

            loginScreenMutableState.value  = loginScreenState.value.copy(
                loginResponseState = loginResponseState)
        }
    }

    private fun saveCurrentUserDetails(login: Login) {
        /* Will call use-case to save user login details to shared preferences WIP */
    }
```